### PR TITLE
initial vulcanized practitionerrole controller

### PR DESCRIPTION
**Pull request by apeirats-va**
This is my first pass at implementing the R4 PractitionerRole controller through Vulcan.

### CODE REVIEW

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  private PractitionerRepository repository;
+
+  /** Vulcan rule for unimplemented parameter. */
+  public static Rule parameterNotImplemented(String parameter) {
+    return (r) -> {
+      if (r.request().getParameter(parameter) != null) {
+        throw new ResourceExceptions.NotImplemented("not-implemented");
+      }
+    };
+  }
```
Comment (apeirats-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:64]: Can probably move method the method above to Vulcan's Rules class. @bschofield-va any opinions?

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+  private VulcanConfiguration<PractitionerEntity> configuration() {
+    return VulcanConfiguration.forEntity(PractitionerEntity.class)
+        .paging(
+            linkProperties.pagingConfiguration(
+                "PractitionerRole", PractitionerEntity.naturalOrder()))
+        .mappings(
+            Mappings.forEntity(PractitionerEntity.class)
+                .value("_id", "cdwId", witnessProtection::toCdwId)
```
Comment (apeirats-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:73]: Need to add a mapping for `practitioner.name`

**Review by jhulbert-va — COMMENTED** [review]
Are the swagger/metadata/IT changes going to be in a separate PR?

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java (hunk 1 above)
Comment (bschofield-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:64]: Do we anticipate needing it more than once? If so, yes.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        .bundling(
+            Bundling.newBundle(PractitionerRole.Bundle::new)
+                .newEntry(PractitionerRole.Entry::new)
+                .linkProperties(linkProperties)
+                .build())
+        .build();
+  }
+
+  /** Validate identifier token. */
+  public boolean tokenIdentifierIsSupported(TokenParameter token) {
```
Comment (jhulbert-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:130]: Can we rename these methods? I think `tokenIdentifier*` is a little bit misleading. Instead, why not something like `practitionerIdentifierTokenIsSupported`? I think that creates better separation between the token itself and the parameter it represents.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+                .token(
+                    "practitioner.identifier",
+                    "npi",
+                    this::tokenIdentifierIsSupported,
+                    this::tokenIdentifierValue)
+                .get())
+        .defaultQuery(returnNothing())
+        .rule(
+            atLeastOneParameterOf(
+                "specialty", "practitioner.identifier", "practitioner.name", "_id"))
```
Comment (jhulbert-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:83]: We should also be supporting the plain old `identifier` param.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+
+  private final LinkProperties linkProperties;
+
+  private PractitionerRepository repository;
+
+  /** Vulcan rule for unimplemented parameter. */
+  public static Rule parameterNotImplemented(String parameter) {
+    return (r) -> {
+      if (r.request().getParameter(parameter) != null) {
+        throw new ResourceExceptions.NotImplemented("not-implemented");
```
Comment (jhulbert-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:61]: This should probably have more information on _what_ parameter isn't implemented.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java (hunk 2 above)
Comment (bschofield-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:73]: I don't know how to interpret this. Does this mean you are still making changes in this PR or this will be implemented in a different PR?

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java (hunk 3 above)
Comment (bschofield-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:130]: because it nicely sorts the tokenXXX methods together. I prefer this style because of the natural grouping of methods.

File: data-query/src/test/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                "http://fonzy.com/r4/PractitionerRole?practitioner.identifier=pr1",
+                9,
+                15));
+    var applied = bundler.apply(vr);
+    assertThat(applied).isEqualTo(expected);
+  }
+
+  @ParameterizedTest
+  @CsvSource({
+    "12345,false",
```
Comment (jhulbert-va) [inline data-query/src/test/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleControllerTest.java:149]: Shouldn't this token format be supported?

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java (hunk 3 above)
Comment (jhulbert-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:130]: Fair enough. I think this should be `tokenPractitionerIdentifier*` still. The plain old `identifier` parameter is also valid, but not a token.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java (hunk 2 above)
Comment (apeirats-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:73]: I'm making/testing the change now and will update the PR.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java (hunk 1 above)
Comment (apeirats-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:64]: The grep for it suggests it's not used anywhere other than PractitionerRole, but I think there may be a case for it in the future.
However I'm happy to just leave it here since, as it stands, it seems to be just a one-of.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java (hunk 3 above)
Comment (apeirats-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:130]: I just added support for a plain `identifier` value that represents just the NPI, and renamed this method to make it clear it represents the `practitioner.identifier` token.

File: data-query/src/test/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleControllerTest.java (hunk 6 above)
Comment (apeirats-va) [reply data-query/src/test/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleControllerTest.java:149]: Followed logic in STU3 implementation, which seems to bail if no system is specified or if the system does not match the exact `us-npi` system.

For the case of R4, if no system is specified should we just assume NPI?

I did add a `identifier` which just takes a NPI value, so that's also an option for the consumer.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java (hunk 2 above)
Comment (jhulbert-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:75]: In the future, GitHub has a draft PR feature if you want to put out a PR while you're still working on features.

File: data-query/src/test/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleControllerTest.java (hunk 6 above)
Comment (jhulbert-va) [reply data-query/src/test/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleControllerTest.java:156]: @jblefkove-va was there a reason for that in the stu3 implementation?

Comment (jblefkove-va) [reply data-query/src/test/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleControllerTest.java:156]: I suspect the STU3 implementation is wrong; it predates our current understanding of token searches.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java (hunk 2 above)
Comment (apeirats-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:70]: Nvm, moved to API-3444

File: data-query/src/test/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleControllerTest.java (hunk 6 above)
Comment (apeirats-va) [reply data-query/src/test/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleControllerTest.java]: Yanked out NPI searches and moved the identifier/name search implementations to API-3444.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+  private VulcanConfiguration<PractitionerEntity> configuration() {
+    return VulcanConfiguration.forEntity(PractitionerEntity.class)
+        .paging(
+            linkProperties.pagingConfiguration(
+                "PractitionerRole", PractitionerEntity.naturalOrder()))
+        .mappings(
+            Mappings.forEntity(PractitionerEntity.class)
+                .value("_id", "cdwId", witnessProtection::toCdwId)
+                .value("identifier", "npi")
+                .string("practitioner.name", "familyName")
```
[the remark this answers is not in the document]
Comment (apeirats-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java]: Moved name search implementation to API-3444

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+ * https://build.fhir.org/ig/HL7/US-Core-R4/StructureDefinition-us-core-practitionerrole.html for
+ * implementation details.
+ */
+@Validated
+@RestController
+@SuppressWarnings("WeakerAccess")
+@RequestMapping(
+    value = "/r4/PractitionerRole",
+    produces = {"application/json", "application/fhir+json"})
+@AllArgsConstructor(onConstructor = @__({@Autowired}))
```
Comment (jhulbert-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:45]: Can you make this `onConstructor_ = @Autowired`? I think it reads better that way.

File: data-query-tests/src/test/java/gov/va/api/health/dataquery/tests/r4/PractitionerRoleIT.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        // search by _id
+        test(
+            200,
+            PractitionerRole.Bundle.class,
+            "PractitionerRole?_id={id}",
+            verifier.ids().practitioner()));
+  }
+
+  @Test
+  public void malformed() {
```
Comment (jhulbert-va) [inline data-query-tests/src/test/java/gov/va/api/health/dataquery/tests/r4/PractitionerRoleIT.java:26]: We don't need the IT's to test the malformed requests. Just each path for reads/searches.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    };
+  }
+
+  private VulcanConfiguration<PractitionerEntity> configuration() {
+    return VulcanConfiguration.forEntity(PractitionerEntity.class)
+        .paging(
+            linkProperties.pagingConfiguration(
+                "PractitionerRole", PractitionerEntity.naturalOrder()))
+        .mappings(
+            Mappings.forEntity(PractitionerEntity.class)
```
Comment (jhulbert-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:69]: What about plain old `identifier`?

**Review by jhulbert-va — COMMENTED** [review]
@bschofield-va @ramosan-lit In regards to the swagger and metadata, I'm not sure whether we want to make that public until the conversation about scopes and scope types is resolved. Thoughts?

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java (hunk 10 above)
Comment (apeirats-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:69]: That's still a token in this case, right? If so, then API-3444

File: data-query-tests/src/test/java/gov/va/api/health/dataquery/tests/r4/PractitionerRoleIT.java (hunk 9 above)
Comment (apeirats-va) [reply data-query-tests/src/test/java/gov/va/api/health/dataquery/tests/r4/PractitionerRoleIT.java:26]: I think I noticed some tests check them, some don't. So I just threw them in but in their own method.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java (hunk 8 above)
Comment (apeirats-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:45]: I could. I just copied it from another controller. I had not realized the syntax you're suggesting was even an option. I tried it out locally and it seems to work fine, so I'll get that committed.

Comment (jhulbert-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:45]: Lombok originally only supported the first way. But the new way is a lot more readable.

File: data-query-tests/src/test/java/gov/va/api/health/dataquery/tests/r4/PractitionerRoleIT.java (hunk 9 above)
Comment (jhulbert-va) [reply data-query-tests/src/test/java/gov/va/api/health/dataquery/tests/r4/PractitionerRoleIT.java:26]: If you want to do that, add the local only assumption. Malformed request tests are unnecessary during a deployment.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java (hunk 10 above)
Comment (jhulbert-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:69]: Cool. That being said, I'll resolve this conversation.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/api/R4DataQueryService.java [hunk 11]
```diff
@@ -114,4 +115,5 @@
         ObservationApi,
         OrganizationApi,
         PatientApi,
+        PractitionerRoleApi,
```
Comment (jhulbert-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/api/R4DataQueryService.java:118]: If all the other search params won't be made available until API-3444. I would prefer we don't update the metadata and swagger right now.

Comment (apeirats-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/api/R4DataQueryService.java:118]: Ok I'll revert that commit.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+                .value("_id", "cdwId", witnessProtection::toCdwId)
+                .get())
+        .defaultQuery(returnNothing())
+        .rule(
+            atLeastOneParameterOf(
+                "_id", "practitioner.name", "practitioner.identifier", "specialty"))
+        .rule(parameterNotImplemented("practitioner.name"))
```
Comment (jhulbert-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:76]: Last thing, lets add plain `identifier` to these rules so if its requested there's a not implemented.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+  private final LinkProperties linkProperties;
+
+  private PractitionerRepository repository;
+
+  /** Vulcan rule for unimplemented parameter. */
+  public static Rule parameterNotImplemented(String parameter) {
+    return (r) -> {
+      if (r.request().getParameter(parameter) != null) {
+        throw new ResourceExceptions.NotImplemented(
+            parameter + " search param is not yet implemented");
```
Comment (bschofield-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:58]: Just a nit, but I'd prefer to avoid casual abbreviations in error messages seen by end users.
_search parameter_ vs _search param_

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java (hunk 12 above)
Comment (bschofield-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:76]: Just a note... all of these rules could be combined as follows and virtually achieve the same result - Disallow these parameters.
```
.rule(atLeastOneParameterOf("_id"))
.rule(forbidUnknownParameter())
```

I don't really understand what returning 503 - Not Implemented is buying us over just returning 400 - Bad Request. We don't specifically 503 all 13 defined query parameters for PractitionerRole. Why just these 3?

Comment (apeirats-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:76]: 501* is Not Implemented.

It's technically a valid request from the user's perspective so returning 400 feels misleading.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java (hunk 13 above)
Comment (apeirats-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:58]: This message only makes its way to our logs and not the response. I'll reword it to `parameter` anyways.

**Comment by jhulbert-va** [comment]
Don't forget the `IcnIncludesMajig`

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java (hunk 12 above)
Comment (bschofield-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:77]: but 400 for the rest these searches is ok? http://hl7.org/fhir/R4/practitionerrole.html#search
I value _consistency_ but will defer to Ramos

Comment (apeirats-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:77]: Hm, I had not even seen that list of parameters TBH.
I based it only on the mandatory ones called out here: http://hl7.org/fhir/us/core/StructureDefinition-us-core-practitionerrole.html#mandatory-search-parameters

I agree with the consistency comment. For the sake of this PR and just making the resource reads available, I don't want to keep flip-flopping on search parameters, but I'm happy to let you make the final call.

API-3444 should finalize the implementation of the required parameters.

Comment (apeirats-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java]: Following conversation on slack, the unsupported search params are now ignored and returning 400.

File: data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+        .paging(
+            linkProperties.pagingConfiguration(
+                "PractitionerRole", PractitionerEntity.naturalOrder()))
+        .mappings(
+            Mappings.forEntity(PractitionerEntity.class)
+                .value("_id", "cdwId", witnessProtection::toCdwId)
+                .get())
+        .defaultQuery(returnNothing())
+        .rule(atLeastOneParameterOf("_id"))
+        .rule(forbidUnknownParameters())
```
Comment (bschofield-va) [inline data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:62]: ^ remove this rule

Comment (apeirats-va) [reply data-query/src/main/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleController.java:62]: Since FHIR says unknown parameters shoudl be simply ignored, should the rule `forbidUnknownParameters` not be used at all by R4 resource controllers?

We had talked about DQ rejecting unknown parameters, but I'm guessing that's no longer the case.

File: data-query/src/test/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleControllerTest.java [hunk 15]
```diff
@@ -0,0 +1,143 @@
...
+import lombok.SneakyThrows;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+public class R4PractitionerRoleControllerTest {
```
Comment (jhulbert-va) [inline data-query/src/test/java/gov/va/api/health/dataquery/service/controller/practitionerrole/R4PractitionerRoleControllerTest.java:37]: ValidRequests test

**Review by jhulbert-va — APPROVED** [review]
